### PR TITLE
IGNITE-16797 DDL support for Storage specific parameters

### PR DESCRIPTION
**Pull request by tkalkirill**
https://issues.apache.org/jira/browse/IGNITE-16797

### CODE REVIEW

File: modules/storage-api/src/main/java/org/apache/ignite/internal/storage/DataStorageManager.java [hunk 1]
```diff
@@ -118,6 +96,73 @@ public Consumer<DataStorageChange> defaultTableDataStorageConsumer(String defaul
         };
     }
 
+    /**
+     * Returns the default data storage.
+     *
+     * @param defaultDataStorageView View of {@link TablesConfigurationSchema#defaultDataStorage}. For the case {@link
+     * UnknownDataStorageConfigurationSchema#UNKNOWN_DATA_STORAGE} and there is only one engine, then it will be the default.
+     */
+    public String defaultDataStorage(String defaultDataStorageView) {
```
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/storage-api/src/main/java/org/apache/ignite/internal/storage/DataStorageManager.java:105]: Discussed personally, we decided to leave it as long as we have the possibility of one storage in the product.

File: modules/storage-api/src/main/java/org/apache/ignite/internal/storage/DataStorageManager.java [hunk 2]
```diff
@@ -118,6 +96,73 @@ public Consumer<DataStorageChange> defaultTableDataStorageConsumer(String defaul
...
+
+                    node.construct(e.getKey(), leafSource, true);
+                }
+            }
+        };
+
+        return tableDataStorageChange -> {
+            tableDataStorageChange.convert(dataStorage);
+
+            configurationSource.descend(((ConstructableTreeNode) tableDataStorageChange));
```
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/storage-api/src/main/java/org/apache/ignite/internal/storage/DataStorageManager.java:162]: Fix it

File: modules/storage-api/src/test/java/org/apache/ignite/internal/storage/DataStorageModulesTest.java [hunk 3]
```diff
@@ -0,0 +1,169 @@
...
+        Map<String, Map<String, Class<?>>> fields = dataStorageModules.collectSchemasFields();
+
+        assertThat(fields, aMapWithSize(2));
+
+        assertThat(fields.get(FIRST), equalTo(Map.of("strVal", String.class, "intVal", int.class)));
+
+        assertThat(fields.get(SECOND), equalTo(Map.of("strVal", String.class, "longVal", long.class)));
+    }
+
+    static DataStorageModule createMockedStorageEngineFactory(
```
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/storage-api/src/test/java/org/apache/ignite/internal/storage/DataStorageModulesTest.java:155]: Fix it

File: modules/storage-api/src/main/java/org/apache/ignite/internal/storage/DataStorageModule.java [hunk 4]
```diff
@@ -15,24 +15,29 @@
...
+     * Returns the unique name of the data storage.
      *
      * <p>Used to map {@link DataStorageConfigurationSchema#name} to {@link StorageEngine}.
      */
     String name();
 
+    /**
+     * Returns the data storage configuration schema.
+     */
+    Class<? extends DataStorageConfigurationSchema> schema();
```
[the remark this answers is not in the document]
Comment (tkalkirill) [reply modules/storage-api/src/main/java/org/apache/ignite/internal/storage/DataStorageModule.java]: Removed this method.

File: modules/storage-api/src/main/java/org/apache/ignite/internal/storage/DataStorageModules.java [hunk 5]
```diff
@@ -112,12 +104,50 @@ public Map<String, StorageEngine> createStorageEngines(
...
-    public Map<String, Map<String, Class<?>>> collectSchemasFields() {
+    public Map<String, Map<String, Class<?>>> collectSchemasFields(Collection<Class<?>> polymorphicSchemaExtensions) {
+        Map<String, Class<? extends DataStorageConfigurationSchema>> schemas = polymorphicSchemaExtensions.stream()
+                .filter(DataStorageConfigurationSchema.class::isAssignableFrom)
+                .collect(toUnmodifiableMap(
+                        schemaCls -> schemaName((Class<? extends DataStorageConfigurationSchema>) schemaCls),
+                        schemaCls -> (Class<? extends DataStorageConfigurationSchema>) schemaCls
+                ));
+
+        Set<String> dataStorageNames = modules.stream()
```
Comment (ibessonov) [inline modules/storage-api/src/main/java/org/apache/ignite/internal/storage/DataStorageModules.java:120]: Ok, so:
modules is a Set that you retrieve by calling "this.modules = modules.values();", it's a proxy.
And after that you convert it back into the map with streams and all this magic. Why can't you store Map as a field?

File: modules/storage-api/src/main/java/org/apache/ignite/internal/storage/DataStorageModules.java [hunk 6]
```diff
@@ -112,12 +104,50 @@ public Map<String, StorageEngine> createStorageEngines(
...
+                .map(DataStorageModule::name)
+                .collect(toSet());
+
+        if (!dataStorageNames.equals(schemas.keySet())) {
+            List<String> dataStorageWithoutSchema = dataStorageNames.stream()
+                    .filter(not(schemas.keySet()::contains))
```
Comment (ibessonov) [inline modules/storage-api/src/main/java/org/apache/ignite/internal/storage/DataStorageModules.java:126]: is there a "containsKey" method? Or simply "contains" maybe?

File: modules/storage-api/src/main/java/org/apache/ignite/internal/storage/DataStorageModules.java [hunk 7]
```diff
@@ -112,12 +104,50 @@ public Map<String, StorageEngine> createStorageEngines(
...
+                        schemaCls -> schemaName((Class<? extends DataStorageConfigurationSchema>) schemaCls),
+                        schemaCls -> (Class<? extends DataStorageConfigurationSchema>) schemaCls
+                ));
+
+        Set<String> dataStorageNames = modules.stream()
+                .map(DataStorageModule::name)
+                .collect(toSet());
+
+        if (!dataStorageNames.equals(schemas.keySet())) {
+            List<String> dataStorageWithoutSchema = dataStorageNames.stream()
```
Comment (ibessonov) [inline modules/storage-api/src/main/java/org/apache/ignite/internal/storage/DataStorageModules.java:125]: Please extract content of this "if" into a method. It is self-contained and pretty big

File: modules/storage-api/src/main/java/org/apache/ignite/internal/storage/DataStorageModules.java (hunk 5 above)
Comment (tkalkirill) [reply modules/storage-api/src/main/java/org/apache/ignite/internal/storage/DataStorageModules.java:120]: Fix it.

File: modules/storage-api/src/main/java/org/apache/ignite/internal/storage/DataStorageModules.java (hunk 6 above)
Comment (tkalkirill) [reply modules/storage-api/src/main/java/org/apache/ignite/internal/storage/DataStorageModules.java:126]: **containsKey** is fine, correct it

File: modules/storage-api/src/main/java/org/apache/ignite/internal/storage/DataStorageModules.java (hunk 7 above)
Comment (tkalkirill) [reply modules/storage-api/src/main/java/org/apache/ignite/internal/storage/DataStorageModules.java:125]: Fix it.

**Pull request merged**